### DISCUSSION TITLE
RPI4 compilation with Code:Bloks IDE

### DISCUSSION THREAD

**Discussion by HDTV101** [raysan5/raylib#1705] [Q&A]
Re-installed RPi4 OS. Followed the raylib instructions to the letter! It seems that the Raylib-3.5.0 must be unzipped into the /tmp directory?

Set up Code Blocks build options to compile and link the shapes_colors_palette.c example.

Compilation and build successful.

Now when I try to run I get an OS error window saying: You must select a host application to "run" a library...

So I changed the build target to GUI application and rebuilt.

Now I get these errors:

||=== Build: Debug in raytest (compiler: GNU GCC Compiler) ===|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlViewport':|
core.c|| undefined reference to `glViewport'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlTextureParameters':|
core.c|| undefined reference to `glBindTexture'|
/usr/bin/ld: core.c|| undefined reference to `glTexParameteri'|
/usr/bin/ld: core.c|| undefined reference to `glTexParameteri'|
/usr/bin/ld: core.c|| undefined reference to `glTexParameteri'|
/usr/bin/ld: core.c|| undefined reference to `glTexParameterf'|
/usr/bin/ld: core.c|| undefined reference to `glTexParameterf'|
/usr/bin/ld: core.c|| undefined reference to `glBindTexture'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlEnableShader':|
core.c|| undefined reference to `glUseProgram'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlDisableShader':|
core.c|| undefined reference to `glUseProgram'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlEnableFramebuffer':|
core.c|| undefined reference to `glBindFramebuffer'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlDisableFramebuffer':|
core.c|| undefined reference to `glBindFramebuffer'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlEnableDepthTest':|
core.c|| undefined reference to `glEnable'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlDisableDepthTest':|
core.c|| undefined reference to `glDisable'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlEnableDepthMask':|
core.c|| undefined reference to `glDepthMask'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlDisableDepthMask':|
core.c|| undefined reference to `glDepthMask'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlEnableBackfaceCulling':|
core.c|| undefined reference to `glEnable'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlDisableBackfaceCulling':|
core.c|| undefined reference to `glDisable'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlEnableScissorTest':|
core.c|| undefined reference to `glEnable'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlDisableScissorTest':|
core.c|| undefined reference to `glDisable'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlScissor':|
core.c|| undefined reference to `glScissor'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlSetLineWidth':|
core.c|| undefined reference to `glLineWidth'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlGetLineWidth':|
core.c|| undefined reference to `glGetFloatv'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlUnloadFramebuffer':|
core.c|| undefined reference to `glBindFramebuffer'|
/usr/bin/ld: core.c|| undefined reference to `glGetFramebufferAttachmentParameteriv'|
/usr/bin/ld: core.c|| undefined reference to `glGetFramebufferAttachmentParameteriv'|
/usr/bin/ld: core.c|| undefined reference to `glDeleteRenderbuffers'|
/usr/bin/ld: core.c|| undefined reference to `glDeleteTextures'|
/usr/bin/ld: core.c|| undefined reference to `glBindFramebuffer'|
/usr/bin/ld: core.c|| undefined reference to `glDeleteFramebuffers'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlClearColor':|
core.c|| undefined reference to `glClearColor'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlClearScreenBuffers':|
core.c|| undefined reference to `glClear'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlUpdateBuffer':|
core.c|| undefined reference to `glBindBuffer'|
/usr/bin/ld: core.c|| undefined reference to `glBufferSubData'|
/usr/bin/ld: ../../../src/libraylib.a(core.o)||in function `rlglInit':|
core.c|| undefined reference to `glGetString'|
/usr/bin/ld: core.c|| undefined reference to `glGetString'|
/usr/bin/ld: core.c|| undefined reference to `glGetString'|
/usr/bin/ld: core.c|| undefined reference to `glGetString'|
/usr/bin/ld: core.c|| undefined reference to `glGetString'|
/usr/bin/ld: core.c|| undefined reference to `eglGetProcAddress'|
/usr/bin/ld: core.c|| undefined reference to `eglGetProcAddress'|
/usr/bin/ld: core.c|| undefined reference to `eglGetProcAddress'|
/usr/bin/ld: core.c|| undefined reference to `glGetFloatv'|
/usr/bin/ld: core.c|| undefined reference to `glDepthFunc'|
/usr/bin/ld: core.c|| undefined reference to `glDisable'|
/usr/bin/ld: core.c|| undefined reference to `glBlendFunc'|
/usr/bin/ld: core.c|| undefined reference to `glEnable'|
/usr/bin/ld: core.c|| undefined reference to `glCullFace'|
/usr/bin/ld: core.c|| undefined reference to `glFrontFace'|
/usr/bin/ld: core.c|| undefined reference to `glEnable'|
||More errors follow but not being shown.|
||Edit the max errors limit in compiler options...|
||=== Build failed: 50 error(s), 0 warning(s) (0 minute(s), 1 second(s)) ===|

**Comment by raysan5** [comment]
This discussion seems a duplicate of a previous discussion. Please, could you post the full compilation line you use?

**Reply by HDTV101** [reply]
I am using the Code Blocks IDE. Note it compiles fine on a compile only. It's when I link or "build" where I get the errors above. I am under the impression RayLib needs no external graphics libraries. Does that mean all that is needed is within libraylib.a?

**Reply by chriscamacho** [reply]
where is code blocks IDE mentioned in the wiki for compiling with PI4 ?

**Reply by HDTV101** [reply]
I did use command line to install Raylib and the examples as per the wiki. But are you saying I cannot use CodeBlocks or Geany for development? That would be odd as even my old Boreland C 3.1 had a reasonable IDE. 

Note as I said above, the examples files do compile. It's the linking that breaks. I mean look at those errors. Where are those files it's complaining about?

The file I am testing with is **shapes_colors_palette.c** . How exactly would you compile and link that so that I get a full screen execution. As I am in native mode, that implies to me it will run as a full screen and drop out of the desktop mode. Is that correct?

**Reply by raysan5** [reply]
Please, could you post the full compilation line you use?

**Reply by HDTV101** [reply]
I am not using a compilation line. I am using the CodeBlocks IDE to edit and compile/link/execute my C code.
I see you are in Barcelona. I don't mean this disrespectfully at all but, do you have a good english understanding? It just seems we are not communicating since I left some detailed information in the posts above.

**Reply by raysan5** [reply]
An IDE is just a frontend, ALL IDEs end-up generating a compilation line, Visual Studio, CodeBlocks and any other IDE in the programming world, that's the base of understanding how code compilation works. I don't use CodeBlocks, so I don't know how to configure it but, if I see the command-line compilation line generated, I can know the issue with your IDE configuration.

**Reply by HDTV101** [reply]
OK thanks,  let me work on that. I have another issue right now where the entire raylib-3.5.0 disappeared from the /tmp directory since the last system reboot? I guess I have to re-install it.

**Reply by HDTV101** [reply]
Yes I learned /tmp is cleared at reboot. Damn, 

So what is your preferred directory path  to extract the source files to? Perhaps this is why I am having these linker issues.

**Reply by HDTV101** [reply]
Is this what you want? Note raytest is the project name

g++  -o bin/Debug/raytest obj/Debug/examples/shapes/shapes_colors_palette.o   

Below are the build or linker errors. The raw compile only of **shapes_colors_palette.c** generates no errors.
**Scroll past this block of errors to see another compile command line and the results**

/usr/bin/ld: obj/Debug/examples/shapes/shapes_colors_palette.o: in function `main':
/home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:23: undefined reference to `InitWindow'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:50: undefined reference to `SetTargetFPS'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:58: undefined reference to `GetMousePosition'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:62: undefined reference to `CheckCollisionPointRec'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:69: undefined reference to `BeginDrawing'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:71: undefined reference to `ClearBackground'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:73: undefined reference to `DrawText'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:74: undefined reference to `GetScreenWidth'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:74: undefined reference to `GetScreenHeight'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:74: undefined reference to `DrawText'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:78: undefined reference to `Fade'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:78: undefined reference to `DrawRectangleRec'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:80: undefined reference to `IsKeyDown'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:82: undefined reference to `DrawRectangle'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:83: undefined reference to `Fade'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:83: undefined reference to `DrawRectangleLinesEx'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:84: undefined reference to `MeasureText'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:84: undefined reference to `DrawText'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:89: undefined reference to `EndDrawing'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:54: undefined reference to `WindowShouldClose'
/usr/bin/ld: /home/pi/raylib-3.5.0/examples/shapes/shapes_colors_palette.c:95: undefined reference to `CloseWindow'
collect2: error: ld returned 1 exit status
Process terminated with status 1 (0 minute(s), 0 second(s))
22 error(s), 0 warning(s) (0 minute(s), 0 second(s))

************************************************************
* I added the path for libraylib.a to the linker options *
************************************************************
g++  -o bin/Debug/raytest obj/Debug/examples/shapes/shapes_colors_palette.o   ../../../src/libraylib.a

Now I get these errors:

/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlViewport':
core.c:(.text+0x75bc): undefined reference to `glViewport'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlTextureParameters':
core.c:(.text+0x9b3c): undefined reference to `glBindTexture'
/usr/bin/ld: core.c:(.text+0x9bb4): undefined reference to `glTexParameteri'
/usr/bin/ld: core.c:(.text+0x9be4): undefined reference to `glTexParameteri'
/usr/bin/ld: core.c:(.text+0x9bfc): undefined reference to `glTexParameteri'
/usr/bin/ld: core.c:(.text+0x9c44): undefined reference to `glTexParameterf'
/usr/bin/ld: core.c:(.text+0x9cb0): undefined reference to `glTexParameterf'
/usr/bin/ld: core.c:(.text+0x9cdc): undefined reference to `glBindTexture'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlEnableShader':
core.c:(.text+0x9d30): undefined reference to `glUseProgram'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlDisableShader':
core.c:(.text+0x9d4c): undefined reference to `glUseProgram'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlEnableFramebuffer':
core.c:(.text+0x9d70): undefined reference to `glBindFramebuffer'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlDisableFramebuffer':
core.c:(.text+0x9d94): undefined reference to `glBindFramebuffer'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlEnableDepthTest':
core.c:(.text+0x9db0): undefined reference to `glEnable'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlDisableDepthTest':
core.c:(.text+0x9dcc): undefined reference to `glDisable'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlEnableDepthMask':
core.c:(.text+0x9de8): undefined reference to `glDepthMask'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlDisableDepthMask':
core.c:(.text+0x9e00): undefined reference to `glDepthMask'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlEnableBackfaceCulling':
core.c:(.text+0x9e18): undefined reference to `glEnable'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlDisableBackfaceCulling':
core.c:(.text+0x9e34): undefined reference to `glDisable'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlEnableScissorTest':
core.c:(.text+0x9e50): undefined reference to `glEnable'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlDisableScissorTest':
core.c:(.text+0x9e6c): undefined reference to `glDisable'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlScissor':
core.c:(.text+0x9ea8): undefined reference to `glScissor'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlSetLineWidth':
core.c:(.text+0x9efc): undefined reference to `glLineWidth'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlGetLineWidth':
core.c:(.text+0x9f2c): undefined reference to `glGetFloatv'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlUnloadFramebuffer':
core.c:(.text+0x9fa4): undefined reference to `glBindFramebuffer'
/usr/bin/ld: core.c:(.text+0x9fb8): undefined reference to `glGetFramebufferAttachmentParameteriv'
/usr/bin/ld: core.c:(.text+0x9fcc): undefined reference to `glGetFramebufferAttachmentParameteriv'
/usr/bin/ld: core.c:(.text+0x9ff4): undefined reference to `glDeleteRenderbuffers'
/usr/bin/ld: core.c:(.text+0xa018): undefined reference to `glDeleteTextures'
/usr/bin/ld: core.c:(.text+0xa024): undefined reference to `glBindFramebuffer'
/usr/bin/ld: core.c:(.text+0xa034): undefined reference to `glDeleteFramebuffers'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlClearColor':
core.c:(.text+0xa120): undefined reference to `glClearColor'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlClearScreenBuffers':
core.c:(.text+0xa140): undefined reference to `glClear'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlUpdateBuffer':
core.c:(.text+0xa170): undefined reference to `glBindBuffer'
/usr/bin/ld: core.c:(.text+0xa184): undefined reference to `glBufferSubData'
/usr/bin/ld: ../../../src/libraylib.a(core.o): in function `rlglInit':
core.c:(.text+0xa1c4): undefined reference to `glGetString'
/usr/bin/ld: core.c:(.text+0xa1e8): undefined reference to `glGetString'
/usr/bin/ld: core.c:(.text+0xa20c): undefined reference to `glGetString'
/usr/bin/ld: core.c:(.text+0xa230): undefined reference to `glGetString'
/usr/bin/ld: core.c:(.text+0xa26c): undefined reference to `glGetString'
/usr/bin/ld: core.c:(.text+0xa3ac): undefined reference to `eglGetProcAddress'
/usr/bin/ld: core.c:(.text+0xa3cc): undefined reference to `eglGetProcAddress'
/usr/bin/ld: core.c:(.text+0xa3ec): undefined reference to `eglGetProcAddress'
/usr/bin/ld: core.c:(.text+0xa818): undefined reference to `glGetFloatv'
/usr/bin/ld: core.c:(.text+0xad08): undefined reference to `glDepthFunc'
/usr/bin/ld: core.c:(.text+0xad10): undefined reference to `glDisable'
/usr/bin/ld: core.c:(.text+0xad1c): undefined reference to `glBlendFunc'
/usr/bin/ld: core.c:(.text+0xad24): undefined reference to `glEnable'
/usr/bin/ld: core.c:(.text+0xad2c): undefined reference to `glCullFace'
/usr/bin/ld: core.c:(.text+0xad34): undefined reference to `glFrontFace'
/usr/bin/ld: core.c:(.text+0xad3c): undefined reference to `glEnable'
Process terminated with status 1 (0 minute(s), 0 second(s))
50 error(s), 0 warning(s) (0 minute(s), 0 second(s))

**Reply by raysan5** [reply]
It seems all required libraries for linking are missing. The following libraries must be added to the linker: -lraylib -lGLESv2 -lEGL -lpthread -lrt -lm -lgbm -ldrm -ldl

**Reply by HDTV101** [reply]
OK Thanks, I'll work on it.

**Reply by HDTV101** [reply]
OK I added the -l files you listed into "other linker options". Except lraylib as that is covered by /src/libraylib.a in the "Link Libraries". 

Fully complies and links without errors.

But when run as a **Console application** I get this in a terminal widow:

INFO: Initializing raylib 3.5
INFO DISPLAY No graphic card set, trying card1
WARNING: DISPLAY: Failed to get DRM resources
INFO: TIMER: Target time per frame: 16.667 milliseconds
INFO: TEXTURE: [ID 0] Unloaded default texture data from VRAM (GPU)
INFO: Window closed sucessfully

Process returned 0 (0x0) execution time : 0.018 s
Press ENTER to continue.

When run as a **GUI applicatio**n no errors but nothing happens it runs in the CodeBlocks debug window area then quits.

**Reply by raysan5** [reply]
Are you compiling it as PLATFORM_DESKTOP? or PLATFORM_DRM?

**Reply by HDTV101** [reply]
I set everything up for PLATFORM_DRM as per your instruction page. I am assuming NATIVE mode means the Raylib application takes over the system until closed. You can't open other windows not part of the Raylib app. In order to open a WEB browser for example, you must close and exit the Raylib app. This is what I want. This is a kiosk type application where the user never sees the operating system.

**Reply by raysan5** [reply]
That's it. Actually, PLATFORM_DRM requires native execution, it does not work if you are the XWindows desktop, it must be executed directly from the command line, before entering the window system. So, if you are using CodeBlocks, you can not execute from there.

**Reply by HDTV101** [reply]
I am getting the same error verbatim using the command line. And I tried building as both a Console application and a GUI application.

INFO: Initializing raylib 3.5
INFO DISPLAY No graphic card set, trying card1
WARNING: DISPLAY: Failed to get DRM resources
INFO: TIMER: Target time per frame: 16.667 milliseconds
INFO: TEXTURE: [ID 0] Unloaded default texture data from VRAM (GPU)
INFO: Window closed sucessfully

**Reply by raysan5** [reply]
That error is probably related to what I mentioned, it fails to get the DRM resource because it's already in use by the XWindow desktop. To execute it you should initialize the RPI as console only (not entering the XWindow system) and execute the program from there. 

If still getting the error, then you should review OpenGL driver configuration on `raspi-config` Advance options, to make sure it is using the KLM driver.

**Reply by HDTV101** [reply]
Where do I find a "how to" to set that up. But this is also going to make it next to impossible to debug? 

So can I simply re-run the commands on your instruction page for the X11 version and just add any needed libraries to my CodeBlocks project?

Perhaps I can use X11 windows for development but the release as console only?

**Reply by raysan5** [reply]
> Where do I find a "how to" to set that up.

I don't know of any tutorial right now, maybe you can Google it...

 > But this is also going to make it next to impossible to debug?

Yes, it would be more difficult than using an IDE but you can always use `gdb` on command line or simply `printf()` traces. Actually that's what I used for raylib development.

**Reply by HDTV101** [reply]
Can I simply re-run the commands on your instruction page for the X11 version and just add any needed libraries to my CodeBlocks project?

Perhaps I can use X11 windows for development but the release as console only?

**Reply by raysan5** [reply]
Yes, to run it over XWindow desktop you need to compile raylib and examples with `PLATFORM_DESKTOP` flag instead of `PLATFORM_DRM`. In that case, you need to link with `-lraylib -lGL -lm -lpthread -ldl -lrt -lX11`.

**Reply by HDTV101** [reply]
But don't I also need to set the system up differently. you instructions say to install one or the other only.

**Reply by raysan5** [reply]
Yes, you can only use one or the other, they are completely different configurations. One runs over XWindow (Desktop) and the other over the kernel (DRM). Internally raylib uses completely different implementations and libraries.

I think that default OpenGL driver configuration on RPI should work for Desktop mode.

**Reply by HDTV101** [reply]
OK it works! Thank you for all your help. As PRi4's are cheap enough I think I'll set another one up to be command line only.

**Reply by raysan5** [reply]
Nice! Glad to know it finally worked! 👍🙂

**Reply by chriscamacho** [reply]
bare in mind the console one, you can set up with network shares and SSH, you can then compile / make on the network share and use ssh to run the executable, but bare in mind input via ssh is different, or you could use a second keyboard attached to the console pi (which would probably be less complication than ssh!)

Good luck with your PI / Rayib adventures !
